model-optimizer/extensions/ops/ExtractImagePatches.py

```diff
@@ -0,0 +1,82 @@
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import numpy as np
+
+from mo.front.common.layout import shape_for_layout, get_batch_dim, get_features_dim
+from mo.front.common.partial_infer.utils import int64_array, tf_window_op_pad_infer
+from mo.graph.graph import Node, Graph
+from mo.ops.op import Op
+
+
+class ExtractImagePatches(Op):
+    op = "ExtractImagePatches"
+
+    def __init__(self, graph: Graph, attrs: dict):
+        assert 'spatial_dims' in attrs, \
+            'ExtractImagePatches operation should have `spatial_dims` parameter set during creation'
+
+        super().__init__(graph, {
+            'type': self.op,
+            'op': self.op,
+            'version': 'opset3',
+            'infer': self.infer,
+            'in_ports_count': 1,
+            'out_ports_count': 1,
+        }, attrs)
+
+    def backend_attrs(self):
+        return [
+            ('sizes', lambda node: ','.join(map(str, node['sizes'][node.spatial_dims]))),
+            ('strides', lambda node: ','.join(map(str, node['strides'][node.spatial_dims]))),
+            ('rates', lambda node: ','.join(map(str, node['rates'][node.spatial_dims]))),
+            'auto_pad',
+        ]
+
+    @staticmethod
+    def infer(node: Node):
+        assert (len(node.in_nodes()) == 1), \
+            'Wrong input nodes number for node {} with type ExtractImagePatches'.format(node.soft_get('name', node.id))
+        input_shape = node.in_port(0).data.get_shape()
+        if input_shape is None:
+            return
+
+        assert len(input_shape) == 4, 'ExtractImagePatches operation supports only 4D tensors'
+
+        layout = node.graph.graph['layout']
+        N = input_shape[get_batch_dim(layout, 4)]
+        C = input_shape[get_features_dim(layout, 4)]
+
+        size_spatial = int64_array(node.sizes)[node.spatial_dims]
+
+        input_spatial_shape = input_shape[node.spatial_dims]
+        stride_spatial_shape = node.strides[node.spatial_dims]
+
+        size_extent = node.rates[node.spatial_dims] * (size_spatial - 1) + 1
+
+        pad_spatial_shape, output_spatial_shape = tf_window_op_pad_infer(input_spatial_shape,
+                                                                         size_extent,
+                                                                         stride_spatial_shape,
+                                                                         node.auto_pad,
+                                                                         False)
+
+        out_shape = shape_for_layout(layout,
```

Comment:
What is layout here? It looks like EIP is always NHWC in TF? Is it correct?

Reply:
We have EIP layer in NCHW while we restore it using MO IR Reader